Update dependecies

Nils-ChristianIseke opened the pull request:
Use noble for CI.

Nils-ChristianIseke committed:
Fix dep (requirement.txt) to versions present in apt

Nils-ChristianIseke committed:
Update Dockerfile Devcontainer.json

Nils-ChristianIseke committed:
Update workflows

Nils-ChristianIseke committed:
Update conf.py

Nils-ChristianIseke committed:
Fix constraints to latest version available via pip

Nils-ChristianIseke committed:
Update python version in workflows.

Nils-ChristianIseke commented on docker/image/Dockerfile:
```diff
@@ -32,6 +33,10 @@ ENV HOME=/home/$user
+RUN . /opt/ros2doc/bin/activate && pip3 install -r requirements.txt -c constraints.txt
+
+RUN echo "source /opt/ros2doc/bin/activate" >> /home/$user/.bashrc
+USER $user
```
Is there any reason that we use a non root user in the container? In the current setup we would need sudo rights for installing pip and apt requirements.

christophebedard replied:
I'm not exactly sure. I looked at the original PR that added it (#1038), and it wasn't discussed. It may be coming from an old Jenkins config, and, given that it's named `rosindex`, I'm guessing it's meant to preserve permissions/owner of the artifact (HTML) files in the [mounted volume/directory](https://build.ros.org/job/doc_ros2doc/1748/consoleFull#console-section-10). I would keep it as-is and use `--user`.

christophebedard replied:
note: `--user` isn't needed since we're now switching to using a venv

Nils-ChristianIseke commented on constraints.txt:
```diff
@@ -1,36 +1,49 @@
-MarkupSafe==2.0.1
-packaging==21.3
-pbr==5.8.0
+alabaster==0.7.16
```
Fixed dependencies to latest available versions because they are working with all tools (didn't worked with the apt version out of the box). Or is there a specific reason to stick to the dep availible via apt?

christophebedard replied:
I think it's OK here if it works, especially since 24.04 is the latest Ubuntu LTS.

christophebedard replied:
however, for traceability, can you provide the commands you ran to bump the versions and update the `constraints.txt` file?

I did this in a `ubuntu:22.04` container after installing `python3-pip` and `python3-venv`:

```console
$ python3 -m venv venv
$ . ./venv/bin/activate
$ pip3 install -r requirements.txt -c constraints.txt
$ pip3 install -U -r requirements.txt
$ pip3 freeze > constraints.txt
```

and the result is a bit different:

```diff
diff --git a/constraints.txt b/constraints.txt
index 1e3e88b3..bbc03a66 100644
--- a/constraints.txt
+++ b/constraints.txt
@@ -1,34 +1,30 @@
-alabaster==0.7.16
-atomicwrites==1.4.1
-attrs==25.3.0
-babel==2.17.0
-certifi==2025.4.26
+alabaster==1.0.0
+Babel==2.14.0
+certifi==2020.6.20
+chardet==4.0.0
 charset-normalizer==3.4.2
-click==8.1.8
+click==8.2.1
 codespell==2.4.1
-doc8==1.1.2
+doc8==2.0.0
 docutils==0.21.2
-idna==3.10
-imagesize==1.4.1
+idna==2.10
+imagesize==1.3.0
 iniconfig==2.1.0
 Jinja2==3.1.6
-MarkupSafe==3.0.2
-more-itertools==10.7.0
+MarkupSafe==2.0.1
 packaging==25.0
-pbr==6.1.1
-pluggy==1.5.0
+pbr==5.8.0
+pluggy==1.6.0
 polib==1.2.0
-py==1.11.0
-Pygments==2.19.1
-pytest==8.3.5
+Pygments==2.17.2
+pyparsing==2.4.7
+pytest==8.4.1
 PyYAML==6.0.2
 regex==2024.11.6
-requests==2.32.3
-restructuredtext_lint==1.4.0
+requests==2.32.4
+restructuredtext_lint==1.3.2
 roman-numerals-py==3.1.0
-setuptools==80.3.1
-six==1.17.0
-snowballstemmer==3.0.0.1
+snowballstemmer==2.2.0
 Sphinx==8.2.3
 sphinx-copybutton==0.5.2
 sphinx-lint==1.0.0
@@ -38,12 +34,11 @@ sphinx-tabs==3.4.7
 sphinx_tamer==0.2.0
 sphinxcontrib-applehelp==2.0.0
 sphinxcontrib-devhelp==2.0.0
-sphinxcontrib-googleanalytics==0.4
 sphinxcontrib-htmlhelp==2.1.0
 sphinxcontrib-jquery==4.1
 sphinxcontrib-jsmath==1.0.1
 sphinxcontrib-mermaid==1.0.0
 sphinxcontrib-qthelp==2.0.0
-sphinxcontrib-serializinghtml==2.0.0
-stevedore==5.4.1
-urllib3==2.4.0
+sphinxcontrib-serializinghtml==1.1.10
+stevedore==3.5.0
+urllib3==1.26.5
```

christophebedard replied:
I updated the `constraints.txt` file and documented the process: https://github.com/ros2/ros2_documentation/pull/5595/commits/fef67ea8d38960699f056eb2e74bfc022a8c1371

Nils-ChristianIseke committed:
Adapt Readme to noble

Nils-ChristianIseke commented:
@christophebedard some open questions from my side. Happy if you could have a look?

christophebedard commented:
I will take a look once I'm back from vacation on Tuesday-ish!

Nils-ChristianIseke commented on docker/image/Dockerfile:
```diff
@@ -3,7 +3,7 @@
 #
 # docker build -f docker/image/Dockerfile .
 
-FROM ubuntu:jammy
+FROM ubuntu:noble
```
Any plans using a deployed Docker image instead of the Dockerfile? I can create an issue for this if you think this is reasonable to do.

christophebedard replied:
What do you mean?

Do you mean installing the dependencies into the image itself instead of installing them through `CMD` when the container actually gets used to build the docs? The way it currently works, the container will install the exact versions of dependencies as specified by the docs/repo being built basically at the same time as the docs are built. I think we should keep it as-is for now, but you can open a follow-up issue.

Nils-ChristianIseke commented on docker/image/Dockerfile:
```diff
@@ -20,7 +20,8 @@ RUN apt-get update && \
         locales \
         make \
-        python3-pip && \
+        python3-pip \
```
Just to support [uv ](https://docs.astral.sh/uv/#projects), we could consider uv instead of pip (much faster, venv support, better dep. management). Can create an issue for this if wanted.

christophebedard replied:
I'd prefer keeping pip. I don't think there's a big enough reason to move to another tool.

christophebedard commented on README.md:
```diff
@@ -37,7 +37,7 @@ pip install -r requirements.txt -c constraints.txt
 ### Pinned versions
 
-For development we currently use Jammy (Ubuntu 22.04) as our build platform.
+For development we currently use Noble (Ubuntu 22.04) as our build platform.
```
```suggestion
For development we currently use Noble (Ubuntu 24.04) as our build platform.
```

christophebedard commented on .github/workflows/test.yml:
```diff
@@ -4,15 +4,15 @@ on: pull_request
         uses: actions/setup-python@v5
         with:
-          python-version: '3.10'
+          python-version: '3.12.2'
```
Can we simply specify `'3.12'` here and everywhere else? Ubuntu 22.04 had [3.10.6](https://packages.ubuntu.com/jammy/python3) and we used `'3.10'`. Ubuntu 24.04 has [3.12.3](https://packages.ubuntu.com/noble/python3), so let's use `'3.12'`.

Nils-ChristianIseke replied:
Done

christophebedard commented on conf.py:
```diff
@@ -109,14 +110,16 @@
+googleanalytics_id = 'G-EVD5Z6G6NH'
+googleanalytics_enabled = True
 # -- Options for HTML output ----------------------------------------------
 
 # The theme to use for HTML and HTML Help pages.  See the documentation for
 # a list of builtin themes.
 #
 html_theme = 'sphinx_rtd_theme'
 html_theme_options = {
-    'analytics_id': 'G-EVD5Z6G6NH',
```
I was wondering why this needed to be moved to `googleanalytics_id` and found the answer here: https://sphinx-rtd-theme.readthedocs.io/en/stable/configuring.html#confval-analytics_id

> **`analytics_id`**
> ...
> _Deprecated since version 3.0.0_: The `analytics_id` option is deprecated, use the [sphinxcontrib-googleanalytics](https://pypi.org/project/sphinxcontrib-googleanalytics/) extension instead.

christophebedard commented on requirements.txt:
```diff
@@ -11,3 +11,4 @@ sphinx-rtd-theme
 sphinx-tabs
 sphinxcontrib-mermaid
 sphinx-tamer
+sphinxcontrib-googleanalytics
```
Small ordering fix:

```suggestion
sphinx-tamer
sphinxcontrib-googleanalytics
sphinxcontrib-mermaid
```

christophebedard commented on .devcontainer/devcontainer.json:
```diff
@@ -1,11 +1,11 @@
 	"build": {
-		"dockerfile": "../docker/image/Dockerfile"
+		"dockerfile": "../docker/image/Dockerfile",
+		"context": ".."
```
The context isn't needed (but the `postCreateCommand` is)  if we keep `CMD`.

christophebedard reviewed (commented):
Sorry for the delay. I have some comments and suggestions.

Nils-ChristianIseke committed:
Update README.md

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>
Signed-off-by: Nils-Christian Iseke <48475933+Nils-ChristianIseke@users.noreply.github.com>

Nils-ChristianIseke committed:
Update requirements.txt

Co-authored-by: Christophe Bedard <bedard.christophe@gmail.com>
Signed-off-by: Nils-Christian Iseke <48475933+Nils-ChristianIseke@users.noreply.github.com>

Nils-ChristianIseke committed:
Use Python 3.12

Signed-off-by: Nils-Christian Iseke <48475933+Nils-ChristianIseke@users.noreply.github.com>

Nils-ChristianIseke committed:
Update devcontainer.json

Signed-off-by: Nils-Christian Iseke <48475933+Nils-ChristianIseke@users.noreply.github.com>

Nils-ChristianIseke committed:
Update devcontainer.json

Signed-off-by: Nils-Christian Iseke <48475933+Nils-ChristianIseke@users.noreply.github.com>

Nils-ChristianIseke commented:
Thank you @christophebedard !

christophebedard commented on docker/image/Dockerfile:
```diff
@@ -32,6 +33,10 @@ ENV HOME=/home/$user
+RUN . /opt/ros2doc/bin/activate && pip3 install -r requirements.txt -c constraints.txt
+
+RUN echo "source /opt/ros2doc/bin/activate" >> /home/$user/.bashrc
+USER $user
```
Have you tried to build and then run the image like it's done by the doc_ros2doc job? https://build.ros.org/job/doc_ros2doc/1752/consoleFull#console-section-10

This is missing a `CMD`, which should now just contain `make multiversion`. However, I don't think adding `"source /opt/ros2doc/bin/activate"` to `/home/$user/.bashrc` works (i.e., it clearly doesn't end up being sourced), so I just added it to the `CMD`:

```dockerfile
CMD ["bash", "-c", "source /opt/ros2doc/bin/activate && make multiversion"]
```

In that case, we don't need to `echo` it to the `~/.bashrc` file. And then I ran into issues with file ownership, as expected:

```
sphinx-multiversion -c . -W  "source" build/html
fatal: detected dubious ownership in repository at '/tmp/doc_repository'
To add an exception for this directory, call:

	git config --global --add safe.directory /tmp/doc_repository
Traceback (most recent call last):
  File "/opt/ros2doc/bin/sphinx-multiversion", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/ros2doc/lib/python3.12/site-packages/sphinx_multiversion/main.py", line 188, in main
    git.get_toplevel_path(cwd=sourcedir_absolute)
  File "/opt/ros2doc/lib/python3.12/site-packages/sphinx_multiversion/git.py", line 25, in get_toplevel_path
    output = subprocess.check_output(cmd, cwd=cwd).decode()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '('git', 'rev-parse', '--show-toplevel')' returned non-zero exit status 128.
make: *** [Makefile:18: multiversion] Error 1
```

so I built the image by providing my own username and uid, which I'm guessing isn't necessary in the doc_ros2doc job because Jenkins is using rosindex and 1000:

```
$ docker build -f docker/image/Dockerfile . -t ros2_documentation_pr5595 --build-arg user=$(whoami) --build-arg uid=$(id -u)
```

then I ran it similar to how the doc_ros2doc job runs it:

```
$ docker run --rm --net=host -v /home/christophe.bedard/ros2_documentation:/tmp/doc_repository -e REPO=/tmp/doc_repository ros2_documentation_pr5595:latest
```

and it seems to work!

Nils-ChristianIseke committed:
Update Dockerfile

Signed-off-by: Nils-Christian Iseke <48475933+Nils-ChristianIseke@users.noreply.github.com>

Nils-ChristianIseke committed:
Update Dockerfile

Signed-off-by: Nils-Christian Iseke <48475933+Nils-ChristianIseke@users.noreply.github.com>

fujitatomoya commented on docker/image/Dockerfile:
```diff
@@ -32,6 +33,9 @@ ENV HOME=/home/$user
 
-USER $user
+CMD ["bash", "-c", "source /opt/ros2doc/bin/activate && make multiversion"]
+RUN python3 -m venv /opt/ros2doc
+COPY requirements.txt constraints.txt .
+RUN . /opt/ros2doc/bin/activate && pip3 install -r requirements.txt -c constraints.txt
```
why do we need to activate the virtual env in the container? this container is already built for ros2 documentation dev environment, what kind of benefit we can bring to the user if we have the virtual env on top of ros2 documentation container image? i do not think we need venv if user is using the container.

christophebedard replied:
so we have to do _something_ here, since `pip` is stricter about installing system-wide packages in Ubuntu 24.04 compared to Ubuntu 22.04. We could use `--break-system-packages` to force it to do it (since it doesn't really matter in the container anyway), but a `venv` just works and is probably the current best practice.

fujitatomoya replied:
ah that is right. i actually removed that https://github.com/ros2/ros2_documentation/pull/5021, venv would be the best option here. thanks!

fujitatomoya commented on docker/image/Dockerfile:
```diff
@@ -32,6 +33,9 @@ ENV HOME=/home/$user
 WORKDIR /tmp/doc_repository
 
-USER $user
+CMD ["bash", "-c", "source /opt/ros2doc/bin/activate && make multiversion"]
```
not related to this PR, but out of curiosity, why do we need to issue `make multiversion` here? `Dockerfile` should be used to create the container image for ros2 documentation development but actually building the documentation during the container creation. i think this should be removed from `Dockerfile` and user should call the make target as they need to do after start the container process.

christophebedard replied:
The `CMD` here (and therefore `make multiversion`) only gets [run when you start the container with this image](https://docs.docker.com/reference/dockerfile/#cmd), not during the image build.

See the commands in my other comment to try it out: https://github.com/ros2/ros2_documentation/pull/5595/files#r2159715095

fujitatomoya replied:
@christophebedard but it is still strange for me to start the `make multiversion` when the user issue `docker run`? my thought is that user should be able to choose the make target in the container, we do not force them to issue `make multiversion`  when the container starts. at least that is what `devcontainer` does.

```console
tomoyafujita@~/DVT/03_OSS/github.com/ros2/ros2_documentation >docker build --rm -f ./docker/image/Dockerfile -t ros2doc:latest .
...
 => => naming to docker.io/library/ros2doc:latest                                                                       0.0s

 tomoyafujita@~/DVT/03_OSS/github.com/ros2/ros2_documentation >docker run -it -v /home/tomoyafujita/DVT/03_OSS/github.com/ros2/ros2_documentation:/tmp/doc_repository ros2doc:latest
 ...
 sphinx-multiversion -c . -W  "source" build/html
outputdir 'foxy' for refs/remotes/origin/foxy conflicts with other versions
outputdir 'humble' for refs/remotes/origin/humble conflicts with other versions
outputdir 'jazzy' for refs/remotes/origin/jazzy conflicts with other versions
outputdir 'kilted' for refs/remotes/origin/kilted conflicts with other versions
outputdir 'rolling' for refs/remotes/origin/rolling conflicts with other versions
Running Sphinx v7.2.6
....
```

this actually takes time to complete the container creation.

right before submitting this comment, all the memory coming back to me. i think we had this discussion before and the reason we have `make multiversion` here under `CMD` in Dockerfile is that documentation pipeline. i do not know the details how documentation pipeline is implemented, but it relies on `make multiversion` in Dockerfile, so we cannot take this out.

christophebedard replied:
yeah, exactly. We could modify the doc_ros2doc job, but it's much simpler if we only modify the Dockerfile here.

Note that we could play with `ENTRYPOINT` + `CMD` so that it sources the venv and then either runs the user-provided command (with `docker run ...`) OR runs `make multiversion` by default:

1. Create this `entrypoint.sh` file alongside the `Dockerfile`:
    ```sh
    #!/bin/bash
    # cd to the mounted docs repo
    cd /tmp/doc_repository
    # source the venv and run the given command
    source /opt/ros2doc/bin/activate
    exec "$@"
    ```
2. In the `Dockerfile`, remove the current `CMD` and replace it with:
    ```dockerfile
    COPY docker/image/entrypoint.sh /opt/ros2doc/entrypoint.sh
    ENTRYPOINT ["/opt/ros2doc/entrypoint.sh"]
    CMD ["make", "multiversion"]
    ```

Build it:

```console
$ docker build -f docker/image/Dockerfile . -t ros2_documentation_pr5595 --build-arg user=$(whoami) --build-arg uid=$(id -u)
```

Then running the image as-is will run `make multiversion` by default (after sourcing the venv):

```console
$ docker run --rm --net=host -v /home/christophe.bedard/ros2_documentation:/tmp/doc_repository -e REPO=/tmp/doc_repository ros2_documentation_pr5595:latest
```

and, if a custom command is provided (e.g., `make html`), it will be run instead of `make multiversion` (after sourcing the venv):

```console
$ docker run --rm --net=host -v /home/christophe.bedard/ros2_documentation:/tmp/doc_repository -e REPO=/tmp/doc_repository ros2_documentation_pr5595:latest make test-tools
python3 -m pytest test/
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-8.3.5, pluggy-1.5.0
rootdir: /tmp/doc_repository
collected 3 items

test/test_macros.py ...                                                  [100%]

============================== 3 passed in 0.03s ===============================
```

@fujitatomoya @Nils-ChristianIseke just an idea. We can consider doing that in another PR.

fujitatomoya reviewed (commented):
thanks for the PR. a couple of comments.

christophebedard commented:
from my end, I think we just need to clarify [how the `constraints.txt` file was updated](https://github.com/ros2/ros2_documentation/pull/5595#discussion_r2160167608), and then we're good.

christophebedard commented:
@Nils-ChristianIseke do you plan to continue working on this PR? The only thing missing is documenting how you generated the `constraints.txt` file: https://github.com/ros2/ros2_documentation/pull/5595#issuecomment-3002225792.

If you don't have time, that's OK, I can take over.

christophebedard committed:
Bump constraints.txt and document how it was generated

Signed-off-by: Christophe Bedard <bedard.christophe@gmail.com>

christophebedard committed:
Document how to use the Dockerfile/image

Signed-off-by: Christophe Bedard <bedard.christophe@gmail.com>

christophebedard commented:
I'm now trying to compare the docs output before and after the bump to Ubuntu 24.04 and the constraints updates.

christophebedard commented:
> I'm now trying to compare the docs output before and after the bump to Ubuntu 24.04 and the constraints updates.

I built the docs in multiversion mode (1) on `rolling` and (2) using this branch (i.e., with the updates `constraints.txt`). I quickly compared some Rolling docs pages, both visually and by `diff`-ing some HTML files. Some non-visible HTML things got moved around, but I think it's functionally the same.

One visual difference I noticed is this, from the end of [this sub-section](https://docs.ros.org/en/rolling/Tutorials/Beginner-CLI-Tools/Recording-And-Playing-Back-Data/Recording-And-Playing-Back-Data.html#record-a-single-topic): `:kbd:` is formatted differently

1. Before:  
    <img width="256" height="45" alt="image" src="https://github.com/user-attachments/assets/3a2d44c4-dac6-4d0d-b2c5-68d4a137b52d" />
2. After:  
    <img width="256" height="45" alt="image" src="https://github.com/user-attachments/assets/acdda4d8-3c0c-450b-8a61-a19425161952" />

I assume it's due to the bump from Sphinx 7.2.6 to 8.2.3. So this LGTM.

christophebedard committed:
Add trailing slash to destination path of COPY

To resolve this error:

  Step 17/19 : COPY requirements.txt constraints.txt .
  When using COPY with more than one source file, the destination must be a directory and end with a /

Signed-off-by: Christophe Bedard <bedard.christophe@gmail.com>

christophebedard commented:
I pushed this branch to this repo (as `christophebedard/Nils-ChristianIseke/UpdateDependecies`), changed the [doc_ros2-documentation job on build.test.ros2.org](https://build.test.ros2.org/job/doc_ros2-documentation/) to point to that branch instead of `rolling` (see [this config diff](https://build.test.ros2.org/job/doc_ros2-documentation/jobConfigHistory/showDiffFiles?timestamp1=2025-04-29_07-56-43&timestamp2=2025-10-06_12-14-58)), and triggered a build.

I had to fix the `COPY` command (b983b7902f7947973dad539f8808cd5be649a427): https://build.test.ros2.org/job/doc_ros2-documentation/2/console. After that, the job passed (minus a mail server connection failure): https://build.test.ros2.org/job/doc_ros2-documentation/3/console. And the resulting docs on the test site look good: https://repo.test.ros2.org/en/rolling/index.html

@claraberendsen I tested this with [build.test.ros2.org/job/doc_ros2-documentation/](https://build.test.ros2.org/job/doc_ros2-documentation/) (24.04) instead of the test version of [build.ros.org/job/doc_ros2doc/](https://build.ros.org/job/doc_ros2doc/) (20.04; there doesn't seem to be a http://build.test.ros.org/). Merging this PR will just make the existing [build.ros.org/job/doc_ros2doc/](https://build.ros.org/job/doc_ros2doc/) job use the updated 24.04 Docker image to build the docs instead of a 22.04 image. Does that sound good to you? It's kind of separate from making the CI job itself use Ubuntu 24.04.

fujitatomoya reviewed (commented):
@Nils-ChristianIseke @christophebedard i think this is breaking the devcontainer environment that i usually use... can you check it on your env too?

christophebedard commented:
I don't use the devcontainer stuff. Can you describe what's broken?

christophebedard committed:
Set Docker build context and update cmd in devcontainer.json

Signed-off-by: Christophe Bedard <bedard.christophe@gmail.com>

christophebedard commented:
Ah, I think it was missing the context, since previously `requirements.txt` and `constraints.txt` weren't `COPY`ied during the image build. And the `postCreateCommand` needed to be updated.

The only other "issue" left is that the venv isn't sourced when you open a terminal, so you need to source it manually (`. /opt/ros2doc/bin/activate`) before building (`make html`). Do you know how to make it source this automatically?

fujitatomoya commented:
i think that we need to rely on the shell in the userland of container, devcontainer does not have control this user environment.

before adding this change, how come do we need the python venv in the 1st place? having docker build environment is kinda dedicated and containerized environment for this specific task? then why do we need python venv??? what is the benefit to have python venv? i maybe missed some context here...

christophebedard committed:
Switch from venv to --break-system-packages

Signed-off-by: Christophe Bedard <bedard.christophe@gmail.com>

christophebedard commented:
I think it was meant to be a cleaner/more modern alternative to using `--break-system-packages` (to get rid of the corresponding error in the Python version shipped with 24.04). And having to install pip dependencies as the user (not simply in a venv) inside the Docker container makes dealing with user/UID a bit more complicated. However, it seems to work, so I guess we can go with that.

christophebedard commented:
But, in my opinion, there [are too many ways to build the docs](https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Contributing-To-ROS-2-Documentation.html):

1. Locally using venv (simplest)
2. GitHub Codespaces
3. Devcontainer
4. Docker image (what's used by CI)

I think the Codespace and devcontainer options are secondary, and I wouldn't necessarily avoid using venvs (and use a more fragile & less modern approach) in the 4th option just to avoid running 1 command in a devcontainer terminal.

fujitatomoya commented:
@christophebedard thanks for fixing up.

i usually use devcontainer to build and test the documentation, Codespaces is actually devcontainer.
AFAIK, we do not really promote using docker directly to the user, right? that means, we actually maintain 2 ways to build the documentation, i say that is not bad ... 🤔 

anyway, lgtm 👍 i think we can address minor adjustment if there are, once we upgrade the base system to ubuntu 24.04. again, thanks for addressing the comments.

christophebedard commented:
I'll run a test doc build/deployment job again (after my last changes) tomorrow just to make sure it's still good.

claraberendsen commented:
> Merging this PR will just make the existing [build.ros.org/job/doc_ros2doc/](https://build.ros.org/job/doc_ros2doc/) job use the updated 24.04 Docker image to build the docs instead of a 22.04 image. Does that sound good to you?

Yes! Thanks for the work here, it's separate but part of the same migration and I was trying to land both so we can push everything to 24.04.

christophebedard commented:
I tested again locally + using a [test job](https://build.test.ros2.org/job/doc_ros2-documentation/5/), it [looks good](https://repo.test.ros2.org/en/rolling/index.html). Let's merge this!

christophebedard commented:
I'm backporting this, because we generally keep the same build setup on all branches. The actual CI job that builds docs only uses the `rolling` branch, but other branches may be used by someone making a single-branch build from the `humble` branch, for example.

christophebedard commented:
I did a quick check. The [doc_ros2doc job on build.ros.org](https://build.ros.org/job/doc_ros2doc/) has been working fine with an Ubuntu 24.04-based image (to build the docs) for a couple of days now, so all good: https://build.ros.org/job/doc_ros2doc/1868/consoleFull#console-section-9

fujitatomoya commented:
@christophebedard thanks for checking that, that is good to know 👍
